plugins/inputs/whois/whois.go

```diff
@@ -0,0 +1,223 @@
+}
+
+func (*Whois) SampleConfig() string {
+	return sampleConfig
+}
+
+func (w *Whois) Gather(acc telegraf.Accumulator) error {
+	now := time.Now()
+
+	if w.Client == nil {
+		if err := w.Init(); err != nil {
+			return err
+		}
+	}
+
+	for _, domain := range w.Domains {
+		w.Log.Debugf("Fetching WHOIS data for: %s", domain)
+		w.Log.Debugf("Using WHOIS server: %s with timeout: %v", w.Server, w.Timeout)
+
+		// Fetch WHOIS raw data
+		rawWhois, err := w.WhoisLookup(w.Client, domain, w.Server)
+		if err != nil {
+			w.Log.Errorf("WHOIS query failed for %s: %v", domain, err)
+			acc.AddError(err)
+
+			// Always register a metric, even on failure
+			acc.AddFields("whois", map[string]interface{}{
+				"status": 0, // Mark failure
+			}, map[string]string{
+				"domain":        domain,
+				"domain_status": "UNKNOWN",
+			})
+			continue
+		}
+
+		// Parse WHOIS data using whois-parser
+		parsedWhois, err := w.ParseWhoisData(rawWhois)
+		if err != nil {
+			w.Log.Errorf("WHOIS parsing failed for %s: %v", domain, err)
+			acc.AddError(err)
+
+			// Always register a metric, even on failure
+			acc.AddFields("whois", map[string]interface{}{
+				"status": 0, // Mark failure
+			}, map[string]string{
+				"domain":        domain,
+				"domain_status": "UNKNOWN",
+			})
+			continue
+		}
+
+		w.Log.Debugf("Parsed WHOIS data for %s: %+v", domain, parsedWhois)
+
+		// Prevent nil pointer panic
+		if parsedWhois.Domain == nil {
+			w.Log.Warnf("No domain info found for %s", domain)
+
+			// Always register a metric, even on failure
+			acc.AddFields("whois", map[string]interface{}{
+				"status": 0, // Mark failure
+			}, map[string]string{
+				"domain":        domain,
+				"domain_status": "UNKNOWN",
+			})
+			continue
+		}
+
+		// Extract expiration date
+		expiration := parsedWhois.Domain.ExpirationDateInTime
+		if expiration == nil {
+			w.Log.Warnf("No expiration date found for %s", domain)
+			continue
+		}
+
+		// Extract creation date
+		created := parsedWhois.Domain.CreatedDateInTime
+		if created == nil {
+			w.Log.Warnf("No created date found for %s", domain)
+			continue
+		}
+
+		// Extract updated date
+		updated := parsedWhois.Domain.UpdatedDateInTime
+		if updated == nil {
+			w.Log.Warnf("No updated date found for %s", domain)
+			continue
+		}
+
+		// Extract DNSSEC status
+		dnssec := parsedWhois.Domain.DNSSec
+
+		// Extract NameServers status
+		nameServers := parsedWhois.Domain.NameServers
+		if len(nameServers) == 0 {
+			w.Log.Warnf("No name servers found for %s", domain)
+			continue
+		}
+
+		// Extract registrar name (handle nil)
+		registrar := ""
+		if parsedWhois.Registrar != nil {
+			registrar = parsedWhois.Registrar.Name
+		}
+
+		// Extract status (handle nil)
+		domainStatus := "UNKNOWN"
+		if parsedWhois.Domain.Status != nil {
+			domainStatus = simplifyStatus(parsedWhois.Domain.Status)
+		}
+
+		// Calculate expiry in seconds
+		expiry := int(expiration.Sub(now).Seconds())
```

Comment:
To prevent wrong data because of some previous lookup(s) did take a long time.
```suggestion
		expiry := int(expiration.Sub(time.Now()).Seconds())
```
Also, not even sure if this field should be added by this plugin. If really needed a user could calculate this with a processor or with a specific query in their TSDB.

Reply:
Here I take the `x509_cert` example, as also have this field. I can remove also it if you don't feel a need, but please let me know.

Reply:
Hmm, I personally have mixed feelings about this, I will let this to the maintainers of Telegraf.

But please apply my suggestion, as that will make the value more accurate.

Reply:
Applied. 👍